### PR TITLE
go-controller: config and docs cleanups, testcase updates, and no logging to file by default

### PR DESCRIPTION
**Pull request by dcbw**
See the individual commit logs for more info on what each one does...

@shettyg @rajatchopra @pecameron @alinbalutoiu

### CODE REVIEW

**Comment by dcbw** [comment]
@alinbalutoiu hopefully I addressed your comments about default values in the other PR here.

**Comment by shettyg** [comment]
Looks like Alin's commit caused a conflict here. I did not think there will be a conflict.

Also, looks like vagrant is failing with this PR with the bootstrapping failing. Let me see if I see anything obvious.

File: docs/config.md [hunk 1]
```diff
@@ -38,37 +45,61 @@ track connections that are initiated from the pods so that the reverse
...
-rundir=
-logdir=
+
+### [ovnsouth] section
+
+This section contains the address and (if the 'ssl' method is used) certificates
+needed to use the OVN southbound database API. Only the the ovn-kubernetes
+master needs to specify the 'server' options.
+```
+address=ssl://1.2.3.4:6641
```
Comment (alinbalutoiu) [inline docs/config.md]: Minor nit, should be 6642 there.

File: etc/ovn_k8s.conf [hunk 2]
```diff
@@ -1,15 +1,35 @@
...
+address=ssl://1.2.3.4:6641
+client-privkey=/path/to/private.key
+client-cert=/path/to/client.crt
+client-cacert=/path/to/client-ca.crt
+server-privkey=/path/to/private.key
+server-cert=/path/to/server.crt
+server-cacert=path/to/server-ca.crt
+
+[ovnsouth]
+address=ssl://1.2.3.4:6641
```
Comment (alinbalutoiu) [inline etc/ovn_k8s.conf]: Minor nit, should be 6642 there.

**Comment by alinbalutoiu** [comment]
The only conflict seems to be at the import step:
```
++<<<<<<< f11fe2d5db764a8e2b24d37dbec6ff5e4063c8ab
 +      "runtime"
++=======
+       "reflect"
++>>>>>>> go-controller: make config defaults clearer and fix OVN address bug
```
Both packages should be there.

Only a couple minor nits, other than that looks good to me.

File: go-controller/pkg/config/config.go [hunk 3]
```diff
@@ -16,15 +17,31 @@ import (
...
-	CNI CNIConfig
+	CNI = CNIConfig{
+		ConfDir: "/etc/cni/net.d",
+		Plugin:  "ovn-k8s-cni-overlay",
+	}
+
 	// Kubernetes holds Kubernetes-related parsed config file parameters and command-line overrides
-	Kubernetes KubernetesConfig
+	Kubernetes = KubernetesConfig{
+		APIServer: "http://localhost:8443",
```
[the remark this answers is not in the document]
Comment (alinbalutoiu) [reply go-controller/pkg/config/config.go]: That should be the default value, I think it should be *http**s**://localhost:8443*

[the remark this answers is not in the document]
Comment (dcbw) [reply go-controller/pkg/config/config.go]: @shettyg @alinbalutoiu yeah, I mean 8080 because we want the unauthenticated port, since by default we don't have/know the certificates necessary for HTTPS.

File: go-controller/pkg/config/config_test.go [hunk 4]
```diff
@@ -161,89 +202,161 @@ var _ = Describe("Config Operations", func() {
...
+					return []string{
+						"address=ssl://1.2.3.4:444",
+						"client-privkey=" + keyFile,
+						"client-cert=/foo/bar/baz",
+						"client-cacert=" + caFile,
+					}
+				})
+
+			generateTests("the SSL scheme cannot bootstrap the CA certificate",
+				"bootstapping /foo/bar/baz CA certificate failed",
```
Comment (shettyg) [inline go-controller/pkg/config/config_test.go]: s/bootstapping/bootstrapping

File: go-controller/pkg/config/config.go [hunk 5]
```diff
@@ -406,32 +424,22 @@ func InitConfig(ctx *cli.Context, defaults *Defaults) error {
...
+		defaults = &Defaults{}
+	}
 
-	if err = validateKubernetesConfig(&cliConfig, &cfg); err != nil {
+	if err = buildKubernetesConfig(&cliConfig, &cfg, defaults); err != nil {
 		return err
 	}
 
-	OvnNorth.ClientAuth, OvnNorth.ServerAuth, err = getOvnAuth(true, &cliConfig.OvnNorth, &cfg.OvnNorth)
+	OvnNorth, err = buildOvnAuth("nb", &cliConfig.OvnNorth, &cfg.OvnNorth, defaults.OvnNorthAddress)
```
Comment (shettyg) [inline go-controller/pkg/config/config.go]: The returned value is not being used anywhere. I am missing something here.

Comment (shettyg) [reply go-controller/pkg/config/config.go]: I also see that the logging from cliconfig is not being copied into config. So any setting to CLI won't work. For e.g., -logfile is not being honored right now.

File: docs/config.md [hunk 6]
```diff
@@ -27,6 +27,13 @@ Copy-Item ".\etc\ovn_k8s.conf" -Destination (New-Item "C:\etc" -Type container -
 
 ## Config values
 
+The config file contains common configuration options shared between the various
```
Comment (pecameron) [inline docs/config.md]: @dcbw - I have started a man page with this material (PR243 docs/ovn_k8s.conf.5)

File: go-controller/README.md [hunk 7]
```diff
@@ -16,42 +16,112 @@ Then find the executables here : go-controller/_output/go/bin/
 
 'ovnkube' is the single all-in-one executable useful for deploying the ovn cni plugin for a kubernetes deployment
 
-'ovn-k8s-cni-overlay' is the cni executable to be planted under /opt/cni/bin so that it can be invoked for each pod event
+'ovn-k8s-cni-overlay' is the cni executable to be placed in /opt/cni/bin so that it can be invoked for each pod event by kubernetes
```
Comment (pecameron) [inline go-controller/README.md]: @dcbw In RHEL/fedora its in /usr/libexec/cni/

File: go-controller/pkg/config/config.go (hunk 5 above)
Comment (dcbw) [reply go-controller/pkg/config/config.go]: > The returned value is not being used anywhere. I am missing something here.

@shettyg OvnNorth is a global defined near the top.

Comment (dcbw) [reply go-controller/pkg/config/config.go]: > I also see that the logging from cliconfig is not being copied into config

@shettyg will fix

**Comment by dcbw** [comment]
@shettyg I've addressed the issue with master/node ordering by splitting things up and making sure the master bits which do set-connection and set-ssl run before the node bits which boostrap the cert.  Also addressed the defaults issue you mention and added testcases to ensure it doesn't happen again.

@pecameron addressed the docs issue.

PTAL thanks!

**Comment by rajatchopra** [comment]
@dcbw gofmt needed on config.go meanwhile.

**Comment by shettyg** [comment]
I think we should set the k8s-apiserver, k8s-token, ovn-remote and ovn-nb even for master.  Right now, we don't really need init-node run on master for it to access other nodes. When we do init-master, we create a OVN logical switch on master and a mgmt port etc. With this change, we can no longer access other nodes from master even though logical switch and mgmt port specific to master node gets created with just init-master.

**Comment by shettyg** [comment]
The other issue is that if you run init-master with SSL, it fails now because the bootstrapping command looks not to be run for just init-master.

**Comment by shettyg** [comment]
Couple of other things.
1. Looks like we are calling ovn-nbctl set-ssl /etc/openvswitch/ovncontroller-privkey.pem /etc/openvswitch/ovncontroller-cert.pem /etc/openvswitch/ovnnb-ca.cert from init-node code too. That is wrong. This probably works at your end because there is ovn-northd running on every node.

2. We were previously only trying to bootstrap using ovn-nbctl and it would work because it supports --bootstrap-ca-cert option. But ovn-sbctl does not. We are currently dependent on ovn-controller to bootstrap ca-cert for us. I think, it makes sense to skip bootstrapping for ovn-sbctl for the time-being.

Basically, the vagrant is broken right now (even without this patch). With this patch, it still does not work either for tcp or ssl. I think the above are the only things to fix.

File: go-controller/pkg/config/config.go [hunk 8]
```diff
@@ -406,32 +456,30 @@ func InitConfig(ctx *cli.Context, defaults *Defaults) error {
...
+		defaults = &Defaults{}
+	}
 
-	CNI.ConfDir = getConfigStr(cliConfig.CNI.ConfDir, cfg.CNI.ConfDir, "/etc/cni/net.d")
-	CNI.Plugin = getConfigStr(cliConfig.CNI.Plugin, cfg.CNI.Plugin, "ovn-k8s-cni-overlay")
+	// Build config that needs no special processing
+	overrideFields(&Default, &cfg.Default)
+	overrideFields(&Default, &cliConfig.Default)
+	overrideFields(&Logging, &cfg.Logging)
+	overrideFields(&Logging, &cliConfig.Logging)
```
Comment (alinbalutoiu) [inline go-controller/pkg/config/config.go]: I don't see where Logging.Level is being honored. In my tests it doesn't matter to which value I change it.
We should follow with a call to logrus.SetLevel for that.
Something like this:
```
func setLogLevel(level int) error {
    switch level {
    case 5:
        logrus.SetLevel(logrus.DebugLevel)
        return nil
    case 4:
        logrus.SetLevel(logrus.InfoLevel)
        return nil
    case 3:
        logrus.SetLevel(logrus.WarnLevel)
        return nil
    case 2:
        logrus.SetLevel(logrus.ErrorLevel)
        return nil
    case 1:
        logrus.SetLevel(logrus.FatalLevel)
        return nil
    case 0:
        logrus.SetLevel(logrus.PanicLevel)
        return nil
    default:
        return fmt.Errorf("setLogLevel: given log level is out of range")
    }
}
```

File: go-controller/pkg/config/config.go [hunk 9]
```diff
@@ -50,14 +67,14 @@ type LoggingConfig struct {
 
 // CNIConfig holds CNI-related parsed config file parameters and command-line overrides
 type CNIConfig struct {
-	// The following config values are used for the CNI plugin.
+	// ConfDir specifies the CNI config directory in which to write the overlay CNI config file
 	ConfDir string `gcfg:"conf-dir"`
-	Plugin  string `gcfg:"plugin"`
+	// ConfDir specifies the name of the CNI plugin
```
Comment (alinbalutoiu) [inline go-controller/pkg/config/config.go]: ```Plugin``` instead of ```ConfDir```

File: go-controller/README.md [hunk 10]
```diff
@@ -16,42 +16,112 @@ Then find the executables here : go-controller/_output/go/bin/
...
-    	comma-separated list of pattern=N settings for file-filtered logging
+     flag to start the central controller that watches pods/services/policies
+  -mtu int
+     MTU value used for the overlay networks (default: 1400)
+  -conntrack-zone int
+     for gateway nodes, the conntrack zone used for conntrack flow rules (default: 64000)
+  -loglevel int
+     log verbosity and level: 5=debug, 4=info, 3=warn, 2=error, 1=fatal (default: 4)
+  -logfile string
+     path of a file to direct log output to (default: /var/log/ovn-kubernetes.log)
```
Comment (alinbalutoiu) [inline go-controller/README.md]: Shouldn't the ```(default: /var/log/ovn-kubernetes.log)``` be removed since by default it doesn't log to a file?

File: go-controller/pkg/config/config.go [hunk 11]
```diff
@@ -406,32 +456,30 @@ func InitConfig(ctx *cli.Context, defaults *Defaults) error {
 		}
 		logrus.Infof("Parsed config file %s", f.Name())
 		logrus.Infof("Parsed config: %+v", cfg)
-
-		if defaults != nil {
-			fillDefaults(&cfg, defaults)
```
Comment (alinbalutoiu) [inline go-controller/pkg/config/config.go]: This was one of the issues that I encountered recently. The defaults values from external_ids were fetched only if the config file was existing, otherwise it wouldn't fetch the values from external_ids. Good to see it fixed here.

File: go-controller/pkg/config/config.go (hunk 8 above)
Comment (dcbw) [reply go-controller/pkg/config/config.go]: @alinbalutoiu it should get called by the individual app, eg ovnkube.go:

```
func runOvnKube(ctx *cli.Context) error {
        if err := config.InitConfig(ctx, nil); err != nil {
                return err
        }

        // Process log flags
        logrus.SetLevel(logrus.Level(config.Logging.Level))
        logrus.SetOutput(os.Stderr)
```

File: go-controller/pkg/config/config.go (hunk 11 above)
Comment (dcbw) [reply go-controller/pkg/config/config.go]: @alinbalutoiu I'd like to do testcases for the external-ids, but we need to bring in something like Kube's FakeExec first.  Will work on that in a follow-up, because I want to unit-test the CA bootstrap functionality too.

File: go-controller/pkg/config/config.go (hunk 8 above)
Comment (alinbalutoiu) [reply go-controller/pkg/config/config.go]: Ah, didn't check ovnkube, this happened to me when using the CNI plugin.
eg ovn-k8s-cni-plugin.go: https://github.com/openvswitch/ovn-kubernetes/blob/master/go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go#L261-L272

Wouldn't it make more sense to do it right after we read the value? This for example also gets ignored since it's debug log and the level will be set only after the function returns. https://github.com/openvswitch/ovn-kubernetes/blob/master/go-controller/pkg/config/config.go#L438-L443

**Comment by dcbw** [comment]
> I think we should set the k8s-apiserver, k8s-token, ovn-remote even for master. Right now, we don't really need init-node run on master for it to access other nodes. When we do init-master, we create a OVN logical switch on master and a mgmt port etc. With this change, we can no longer access other nodes from master even though logical switch and mgmt port specific to master node gets created with just init-master (ovn-controller needs ovn-remote set for it to program master ovs).

@shettyg agreed, will do.

> The other issue is that if you run init-master with SSL, it fails now because the bootstrapping command looks not to be run for just init-master. You likely need to call config.*.ServerAuth.ConfigureAuth() from master too. I guess, it is okay if it is called twice when someone does 'init-master' and 'init-node' at the same time (I have never run them on the same node though)

But isn't the master the thing that actually *sets* the SSL certs for the database?  I mean, the master is what configures the database connection method, so it should already have the CA cert specified in CLI or config file.  Before the master runs, the database will not have SSL set up, and something has to provide the CA when doing set-ssl and that's currentl the master.  Or am I not quite getting how the DB gets configured?

> Couple of other things.
> 
>     Looks like we are calling ovn-nbctl set-ssl /etc/openvswitch/ovncontroller-privkey.pem /etc/openvswitch/ovncontroller-cert.pem /etc/openvswitch/ovnnb-ca.cert from init-node code too. That is wrong. This probably works at your end because there is ovn-northd running on every node.

Yep, that's wrong if that's happening.

>     We were previously only trying to bootstrap using ovn-nbctl and it would work because it supports --bootstrap-ca-cert option. But ovn-sbctl does not. We are currently dependent on ovn-controller to bootstrap ca-cert for us. I think, it makes sense to skip bootstrapping for ovn-sbctl for the time-being.
> 
> Basically, the vagrant is broken right now (even without this patch). With this patch, it still does not work either for tcp or ssl. I think the above are the only things to fix.

**Comment by dcbw** [comment]
@shettyg replying to myself on this, but sbctl does not in fact support --bootstrap-ca-cert or --peer-ca-cert.  Removed those in https://github.com/openvswitch/ovs/pull/220

**Comment by shettyg** [comment]
>But isn't the master the thing that actually sets the SSL certs for the database? I mean, the master is what configures the database connection method, so it should already have the CA cert specified in CLI or config file. Before the master runs, the database will not have SSL set up, and something has to provide the CA when doing set-ssl and that's currentl the master. Or am I not quite getting how the DB gets configured?

The master uses the set-ssl method to configure the server certificates. I.e certificates that ovsdb-server backing ovn-nb and ovn-sb should be using to accept new connections. It also sets the IP address to use to accept ssl connections only.  Then the code goes ahead and tries to do a 'init-master' which gets into the client mode to configure the databases via ovn-nbctl. Since the servers only accept ssl connections (or unix domain sockets), and ovn-nbctl is being configured to run using ssl in our code, it needs client certificates to connect (which needs bootstrapping).  We can change ovn-nbctl to use unix domain sockets too - and I agree that people can supply the ca cert to use instead of bootstrapping. But if they don't, then it fails. I think we need to rationalize it, but may we can do it in a future commit instead of overloading many things here

 >Are you sure? I only did this for sbctl because its manpage shows it supports --bootstrap-ca-cert.

yeah, I saw that. There is a bug in ovn-sbctl where it does not accept --bootstrap-ca-cert option. Any fix in OVN itself, means that this code will remain broken for longer.

**Comment by shettyg** [comment]
If the current code supports ovn-nbctl to use unix domain socket when running in master, I think, we can agree that being the right method. The last time I looked, it was broken - maybe it is fixed now with your recent commit?

**Comment by dcbw** [comment]
@shettyg @alinbalutoiu @rajatchopra updated to address all comments

But can one of you test vagrant?  I cannot bring up vagrant with SSL because it requires a shared directory between all the machines for the cert request signing, and the libvirt vagrant config defaults to rsync, not NFS.  And changing that to NFS leads down a rathole of NFS which I don't quite have time to figure out tonight...

**Comment by shettyg** [comment]
TCP works. SSL fails - but that is just due to the fact that we are trying to bootstrap ovn-sbctl - I guess, you expected ovn-controller to take care of it before calling ovnkube (I was expecting that we will skip ovn-sbctl bootstrap and let ovn-controller restart by ovnkube bootstrap it.). But that is okay. I will send an incremental to likely just error message bootstrap error instead of returning error. and let the first actual use of the utility fail.
